### PR TITLE
[MDS-519] Carousel v2 with intersection observer

### PR DESCRIPTION
**Pull request by tkullisaar**

### CODE REVIEW

File: assets/js/hooks/carousel.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                behavior: "smooth",
+                inline: "nearest",
+                boundary: this.reel,
+            });
+        }
+    },
+
+    initScrolls() {
+        if (document && "scrollBehavior" in document.documentElement.style) {
+        this.scrollIntoViewSmoothly = scrollIntoView;
```
Comment (karl-kallavus) [inline assets/js/hooks/carousel.js:51]: minor: identation

File: assets/js/hooks/carousel.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    
+    handleLeftArrowClick() {
+        const firstVisibleIndex = this.findFirstVisibleIndex();
+        const step = parseInt(this.el.dataset.step, 10);
+        const actualScrollForIndex = firstVisibleIndex < step 
+            ? 0 
+            : firstVisibleIndex - step;
+        const scrollToIndex = this.itemsArray.findIndex((item, index) => index === actualScrollForIndex);
+        
+            if (scrollToIndex !== -1) {
```
Comment (karl-kallavus) [inline assets/js/hooks/carousel.js:65]: identation

File: assets/js/hooks/carousel.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    handleRightArrowClick() {
+        const lastVisibleIndex = this.findLastVisibleIndex();
+        const step = parseInt(this.el.dataset.step, 10);
+        const lastIndex = this.itemsArray.length - 1;
+        const actualScrollForIndex = lastIndex - lastVisibleIndex < step
+            ? lastIndex
+            : lastVisibleIndex + step;
+        const scrollToIndex = this.itemsArray.findIndex((item, index) => index === actualScrollForIndex);
+        
+            if (scrollToIndex !== -1) {
```
Comment (karl-kallavus) [inline assets/js/hooks/carousel.js:84]: this and line 65 could be a function

File: assets/js/hooks/carousel.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    getArray(selector, element) {
+        const elements = Array.from(element.querySelectorAll(selector));
+        return elements;
+    },
+
+    initIntersectionObserver() {
+        const observer = new IntersectionObserver(
+            (entries) => {
+              entries.forEach((entry) => {
+                entry.intersectionRatio >= 0.75
```
Comment (karl-kallavus) [inline assets/js/hooks/carousel.js:116]: make a constant for 0.75

**Review by karl-kallavus — CHANGES_REQUESTED** [review]
Please install `prettier`. I see only tine indentation knit picks and minor refactoring.

File: assets/js/hooks/carousel.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                block: "nearest",
+                behavior: "smooth",
+                inline: "nearest",
+                boundary: this.reel,
+            });
+        }
+    },
+
+    initScrolls() {
+        if (document && "scrollBehavior" in document.documentElement.style) {
```
Comment (karl-kallavus) [inline assets/js/hooks/carousel.js:50]: NB: I need to double check how it actually works. I think we will discuss it on our call @tkullisaar

**Review by phgrey — APPROVED** [review]

**Review by karl-kallavus — APPROVED** [review]
LGTM

**Comment by karl-kallavus** [comment]
Still we need to fix scroll-snap position thought

File: lib/moon/design/carousel/reel.ex [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  def render(assigns) do
+    ~F"""
+    <ul
+      {=@id}
+      class={merge([
+        [
+          "flex overflow-x-auto overflow-y-hidden h-auto hidden-scroll gap-4 no-scrollbar",
+          "[-webkit-overflow-scrolling:touch] [scrollbar-width:none]",
+          "[-ms-overflow-style:-ms-autohiding-scrollbar]",
+          "[&>li]:list-none before:absolute [&>li]:before:content-[&quot;200B&quot;]",
```
Comment (karl-kallavus) [inline lib/moon/design/carousel/reel.ex:51]: **[&>li]:** before:absolute

File: lib/moon/design/carousel/reel.ex [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+          "[&>*]:flex-[0_0_auto] [&>img]:h-full [&>img]:basis-auto [&>img]:w-auto",
+          "snap-x rtl:flex-row-reverse",
```
Comment (karl-kallavus) [inline lib/moon/design/carousel/reel.ex:53]: `snap-mandatory` is missing, I assume that might be the case

**Pull request merged**